engine/rpcserver.go

```diff
@@ -1213,3 +1221,322 @@ func (s *RPCServer) GetAuditEvent(ctx context.Context, r *gctrpc.GetAuditEventRe
+		Data:   gctVM.ShortName() + " (" + gctVM.ID.String() + ") executed",
+	}, nil
+}
+
+// GCTScriptStop terminate a running script
+func (s *RPCServer) GCTScriptStop(ctx context.Context, r *gctrpc.GCTScriptStopRequest) (*gctrpc.GCTScriptGenericResponse, error) {
+	if !gctscript.GCTScriptConfig.Enabled {
+		return &gctrpc.GCTScriptGenericResponse{Status: gctscript.ErrScriptingDisabled.Error()}, nil
+	}
+
+	UUID, err := uuid.FromString(r.Script.UUID)
+	if err != nil {
+		return &gctrpc.GCTScriptGenericResponse{Status: "error", Data: err.Error()}, nil
+	}
+
+	if v, f := gctscript.AllVMs[UUID]; f {
+		err = v.Shutdown()
+		status := " terminated"
+		if err != nil {
+			status = err.Error()
+		}
+		return &gctrpc.GCTScriptGenericResponse{Status: "ok", Data: v.ID.String() + status}, nil
+	}
+	return &gctrpc.GCTScriptGenericResponse{Status: "error", Data: "no running script found"}, nil
+}
+
+// GCTScriptUpload upload a new script to ScriptPath
+func (s *RPCServer) GCTScriptUpload(ctx context.Context, r *gctrpc.GCTScriptUploadRequest) (*gctrpc.GCTScriptGenericResponse, error) {
+	if !gctscript.GCTScriptConfig.Enabled {
+		return &gctrpc.GCTScriptGenericResponse{Status: gctscript.ErrScriptingDisabled.Error()}, nil
+	}
+
+	fPath := filepath.Join(gctscript.ScriptPath, r.ScriptName)
+	_, err := os.Stat(fPath)
+	if err != nil {
+		if !os.IsNotExist(err) {
+			return nil, err
+		}
+	} else {
+		if !r.Overwrite {
+			return nil, fmt.Errorf("%s script found and overwrite set to false", r.ScriptName)
+		} else if r.Overwrite {
+			f := filepath.Join(gctscript.ScriptPath, "version_history")
+			err = os.MkdirAll(f, 0770)
+			if err != nil {
+				return nil, err
+			}
+			timeString := strconv.FormatInt(time.Now().UnixNano(), 10)
+			renamedFile := filepath.Join(f, timeString+"-"+r.ScriptName)
+			err = file.Move(fPath, renamedFile)
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	newFile, err := os.Create(fPath)
+	if err != nil {
+		return nil, err
+	}
+	defer newFile.Close()
+
+	_, err = newFile.Write(r.Data)
+	if err != nil {
+		return nil, err
+	}
+	err = newFile.Close()
+	if err != nil{
+		log.Errorln(log.Global, "failed to close file handle, archive removal may fail")
+	}
+
+	if r.Archived {
+		z, errZip := zip.OpenReader(fPath)
+		if errZip != nil {
+			return nil, err
+		}
+
+		defer z.Close()
+
+		for x := range z.Reader.File {
+			var zFile io.ReadCloser
+			zFile, err = z.Reader.File[x].Open()
+			if err != nil {
+				return nil, err
+			}
+
+			zipPath := filepath.Join(gctscript.ScriptPath, z.Reader.File[x].Name) // nolint:gosec
+			// We ignore gosec linter above because the code below files the file traversal bug when extracting archives
+			if !strings.HasPrefix(zipPath, filepath.Clean(gctscript.ScriptPath)+string(os.PathSeparator)) {
+				_ = zFile.Close()
+				return nil, fmt.Errorf("%s: illegal file path", fPath)
+			}
+
+			if z.Reader.File[x].FileInfo().IsDir() {
+				err = os.MkdirAll(zipPath, z.Reader.File[x].Mode())
```

Comment:
zipPath at this point would be a file instead of a directory

Reply:
This is to maintain file path inside the zip when extracting

Reply:
Yeah I saw what you were trying to do but on Windows this would create a directory called "testdir/test123.gct/" (assuming the archive has a directory inside it called testdir and a file in that directory called test123.gct) if IsDir() did actually work

Reply:
I am unable to replicate this on a windows 10 pro machine

```powershell
PS D:\gct> .\gctcli.exe gctscript upload ".\gctscript\archived.zip" --archived=true
{
 "status": "ok",
 "data": "script C:\\Users\\x\\AppData\\Roaming\\GoCryptoTrader\\scripts\\archived.zip written"
}
PS D:\gct>

PS D:\gct> dir C:\\Users\\x\\AppData\\Roaming\\GoCryptoTrader\\scripts\\


    Directory: C:\Users\x\AppData\Roaming\GoCryptoTrader\scripts


Mode                LastWriteTime         Length Name
----                -------------         ------ ----
d-----        6/01/2020   6:14 PM                archived
d-----        6/01/2020   6:14 PM                version_history
-a----        2/01/2020  11:24 PM            130 2.gct
-a----        2/01/2020  11:22 PM            186 orderbook.gct
-a----        2/01/2020  10:33 PM            130 timer.gct


PS D:\gct> dir C:\\Users\\x\\AppData\\Roaming\\GoCryptoTrader\\scripts\\archived


    Directory: C:\Users\x\AppData\Roaming\GoCryptoTrader\scripts\archived


Mode                LastWriteTime         Length Name
----                -------------         ------ ----
-a----        6/01/2020   6:16 PM             40 one.gct
-a----        6/01/2020   6:16 PM             40 two.gct


PS D:\gct>
```

Also with with folders inside the archive as well

```powershell
PS D:\gct> Get-ChildItem -Recurse C:\Users\x\AppData\Roaming\GoCryptoTrader\scripts\archive_with_folder\


    Directory: C:\Users\x\AppData\Roaming\GoCryptoTrader\scripts\archive_with_folder


Mode                LastWriteTime         Length Name
----                -------------         ------ ----
d-----        6/01/2020   6:19 PM                test
-a----        6/01/2020   6:19 PM             49 once.gct
-a----        6/01/2020   6:19 PM            130 timer.gct


    Directory: C:\Users\x\AppData\Roaming\GoCryptoTrader\scripts\archive_with_folder\test


Mode                LastWriteTime         Length Name
----                -------------         ------ ----
-a----        6/01/2020   6:19 PM             49 once.gct
-a----        6/01/2020   6:19 PM            130 timer.gct


PS D:\gct>
```

extracts to expected locations i will need to investigate further

Reply:
On commit: 567df2ace38fd2bfa514e836e33a12f5742c59de

```
gctcli.exe gctscript upload --path=C:\GCTScripts\testscript.zip --overwrite --archived
2020/01/07 08:45:31 rpc error: code = Unknown desc = open C:\Users\X\AppData\Roaming\GoCryptoTrader\scripts\testdir\test123.gct: The system cannot find the path specified.
```

Failing here: https://github.com/thrasher-corp/gocryptotrader/pull/383/files#diff-9a26a62ed888d995d75ee011e28aa4e6R1409

z.Reader.File[x].FileInfo().IsDir() returns always false on my PC and no directory path is ever supplied when iterating over z.Reader.File, only files

archive is setup as:

```
  testscript.zip
      - testdir\test123.gct
      - pairs.gct
      - test.gct
```

Reply:
I get the same if I delete the contents of the scripts dir, essentially starting fresh:

```
gctcli.exe gctscript upload --path=C:\GCTScripts\testscript.zip --archived
2020/01/07 08:53:13 rpc error: code = Unknown desc = open C:\Users\X\AppData\Roaming\GoCryptoTrader\scripts\testdir\test123.gct: The system cannot find the path specified.
```

Reply:
A fix has been pushed out that addresses this zip files now extract correctly and by default they will extract info a folder with the same name as the zip file :D